File: api/callback.js
export default async function handler(req, res) {
  // Set CORS headers
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

  if (req.method === 'OPTIONS') {
    res.status(200).end();
    return;
  }

  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  try {
    const { code, redirect_uri, code_verifier } = req.body;

    console.log('Received data:', { code, redirect_uri, code_verifier });

    if (!code || !redirect_uri || !code_verifier) {
      console.log('Missing required fields');
      return res.status(400).json({
        error: 'Missing required fields',
        received: req.body
      });
    }

    // Check if Spotify credentials are configured (fallback to frontend's public client_id)
    const client_id = process.env.SPOTIFY_CLIENT_ID || '2ee0d98b21d048978bf73d78924daf91';
    console.log('Using client_id:', client_id);

    const tokenUrl = 'https://accounts.spotify.com/api/token';
    const payload = new URLSearchParams({
      client_id: client_id,
      grant_type: 'authorization_code',
      code: code,
      redirect_uri: redirect_uri,
      code_verifier: code_verifier,
    });

    console.log('Sending to Spotify:', Object.fromEntries(payload));

    const response = await fetch(tokenUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: payload,
    });

    console.log('Spotify response status:', response.status);
    const responseText = await response.text();
    console.log('Spotify response:', responseText);

    if (!response.ok) {
      const errorData = {
        error: 'Token exchange failed',
        status: response.status,
        response: responseText
      };
      
      try {
        errorData.details = JSON.parse(responseText);
      } catch {
        errorData.details = responseText;
      }
      
      return res.status(500).json(errorData);
    }

    const spotifyResponse = JSON.parse(responseText);
    console.log('Successfully got tokens:', Object.keys(spotifyResponse));
    
    return res.status(200).json(spotifyResponse);
  } catch (error) {
    console.error('Error in callback:', error);
    return res.status(500).json({
      error: 'Internal server error',
      details: error.message
    });
  }
}

File: api/callback.py
from flask import Flask, request, jsonify
import os
import requests
from dotenv import load_dotenv

load_dotenv()

app = Flask(__name__)

@app.route('/api/callback', methods=['POST'])
def callback():
    try:
        data = request.get_json()
        print(f"Received data: {data}")
        
        if not data:
            print("No JSON data received")
            return jsonify({'error': 'No JSON data received'}), 400
        
        code = data.get('code')
        redirect_uri = data.get('redirect_uri')
        code_verifier = data.get('code_verifier')

        print(f"Code: {code}")
        print(f"Redirect URI: {redirect_uri}")
        print(f"Code verifier: {code_verifier}")

        if not code or not redirect_uri or not code_verifier:
            print("Missing required fields")
            return jsonify({'error': 'Missing required fields', 'received': data}), 400

        # Check if Spotify credentials are configured (fallback to frontend's public client_id for local dev)
        client_id = os.environ.get('SPOTIFY_CLIENT_ID') or '2ee0d98b21d048978bf73d78924daf91'
        print(f"Using client_id: {client_id}")

        token_url = 'https://accounts.spotify.com/api/token'
        payload = {
            'client_id': client_id,
            'grant_type': 'authorization_code',
            'code': code,
            'redirect_uri': redirect_uri,
            'code_verifier': code_verifier,
        }

        headers = {
            'Content-Type': 'application/x-www-form-urlencoded'
        }

        print(f"Sending to Spotify: {payload}")
        response = requests.post(token_url, data=payload, headers=headers)
        print(f"Spotify response status: {response.status_code}")
        print(f"Spotify response: {response.text}")

        if response.status_code != 200:
            error_data = {'error': 'Token exchange failed', 'status': response.status_code, 'response': response.text}
            try:
                error_data['details'] = response.json()
            except:
                error_data['details'] = response.text
            return jsonify(error_data), 500

        spotify_response = response.json()
        print(f"Successfully got tokens: {list(spotify_response.keys())}")
        return jsonify(spotify_response)
    except Exception as e:
        print(f"Error in callback: {str(e)}")
        import traceback
        traceback.print_exc()
        return jsonify({'error': 'Internal server error', 'details': str(e)}), 500

def handler(request):
    return app(request.environ, request.start_response)